sdk/core/Azure.Core.Extensions/src/ClientInformation.cs

```diff
@@ -0,0 +1,56 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Runtime.ExceptionServices;
+
+namespace Azure.Core.Extensions
+{
+    internal class ClientRegistration<TClient, TOptions>
+    {
+        public string Name { get; }
+
+        private readonly Func<TOptions, TClient> _factory;
+
+        private TClient _cachedClient;
+
+        private ExceptionDispatchInfo _cachedException;
+
+        public ClientRegistration(string name, Func<TOptions, TClient> factory)
+        {
+            Name = name;
+            _factory = factory;
+        }
+
+        public TClient GetClient(TOptions options)
+        {
+            _cachedException?.Throw();
+
+            if (_cachedClient != null)
+            {
+                return _cachedClient;
+            }
+
+            lock (this)
```

Comment:
Should we consider locking against a sentinel object rather than the instance?   I know this class is marked for internal use, but unless there's a specific reason for `this` that I'm not privy to, it may be a good idea to err on the side of caution.

Reply:
Done